### PR TITLE
Basic PN-MCTS in the full Ludii repo

### PR DESCRIPTION
**Pull request by DennisSoemers**
This should be it.

### CODE REVIEW

File: AI/src/search/mcts/MCTS.java [hunk 1]
```diff
@@ -593,6 +594,8 @@ public static MCTS createPVTS(final Features features, final Heuristics heuristi
 							
 							while (current.contextRef().trial().status() == null)
 							{
+								// TODO should break early for proven nodes and backpropagate proven value
```
Comment (DennisSoemers) [inline AI/src/search/mcts/MCTS.java:597]: I still did not add that, if we select a node that is already proven, we can just break out of tree traversal and skip play-out and immediately backpropagate. Think I didn't see it in your code either, but it seems like an obvious thing to add.

File: AI/src/search/mcts/finalmoveselection/RobustChild.java [hunk 2]
```diff
@@ -15,6 +15,9 @@
  * with an additional tie-breaker based on value estimates. If the MCTS
  * has a learned selection policy, that can be used as a second tie-breaker.
  * 
+ * EXCEPTION: if the child with the best expected value (*not* visits) has its
+ * value proven, we will prioritise that child.
+ *
```
Comment (DennisSoemers) [inline AI/src/search/mcts/finalmoveselection/RobustChild.java:20]: for now, I updated RobustChild as follows: if our child with the maximum score (over all children) happens to have its score proven, we just select that child, even if it doesn't have the highest visit count. This should also work with my ScoreBounded implementation, not only PN-MCTS.

I didn't implement anything with contempt factor.

I do think there's a funny thing about my implementation though. Suppose we have three children:
- A: avg value = -0.1, visits = 1000, not proven
- B: avg value = 0.0, visits = 500, proven (so the value of 0.0 is proven)
- C: avg value = 0.1, visits = 800, not proven

In this case, my implementation will now choose child C. Actually, only if the children are sorted as A, B, C. Then A gets kicked out first because there exists a higher proven value (B), but then gets overridden by C because C has an even higher value (even though not proven) + higher visits, but I no longer compare visits to A. If I have the same children just ordered differently, I might still pick A.... this seems bad :D

File: AI/src/search/mcts/nodes/PNMCTSNode.java [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+	protected PNMCTSNodeTypes type;
+	
+	/** The value (in terms of proven/disproven/dont know) for this node */
+	protected PNMCTSNodeValues proofValue;
+	
+	/** Are the cached PNS-based terms of childrens' selection scores outdated? */
+	protected boolean childSelectionScoresDirty = false;
+	
+	/** Cached PNS-based terms for selection scores for all our children (including unexpanded ones) */
+	protected final double[] childrenPNSSelectionTerms;
```
Comment (DennisSoemers) [inline AI/src/search/mcts/nodes/PNMCTSNode.java:69]: I stored all these terms for all children together in the parent node.  This lets me also store PNS-terms for children that do not exist yet.

File: AI/src/search/mcts/nodes/PNMCTSNode.java [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+    
+    /**
+     * Evaluates a node as in PNS according to L. V. Allis' 
+     * "Searching for Solutions in Games and Artificial Intelligence"
+     */
+    public void evaluate() 
+    {
+        if (this.context.trial().over()) 
+        {
+            if (RankUtils.utilities(this.context)[rootPlayer] == 1.0)
```
Comment (DennisSoemers) [inline AI/src/search/mcts/nodes/PNMCTSNode.java:128]: this 1.0 would have to change if we want to be able to prove things in games with more than 2 players

File: AI/src/search/mcts/nodes/PNMCTSNode.java [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+						if (childNode.disproofNumber < disproof)
+						{
+							disproof = childNode.disproofNumber;
+						}
+					}
+					else
+					{
+						// An unexpanded child
+						// TODO verify that this is indeed what we want to do?
+						// NOTE: addition for proof, setting for disproof
```
Comment (DennisSoemers) [inline AI/src/search/mcts/nodes/PNMCTSNode.java:179]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/nodes/PNMCTSNode.java [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+						if (childNode.proofNumber < proof)
+						{
+							proof = childNode.proofNumber;
+						}
+					}
+					else
+					{
+						// An unexpanded child
+						// TODO verify that this is indeed what we want to do?
+						// NOTE: addition for disproof, setting for proof
```
Comment (DennisSoemers) [inline AI/src/search/mcts/nodes/PNMCTSNode.java:214]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/nodes/PNMCTSNode.java [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+    
+    /**
+     * Store a flag saying whether the cached PNS-based terms of our childrens'
+     * selection scores are (potentially) outdated.
+     * 
+     * @param newFlag
+     */
+    public void setSelectionScoresDirtyFlag(final boolean newFlag)
+    {
+    	childSelectionScoresDirty = newFlag;	// TODO selection strategy will have to account for this
```
Comment (DennisSoemers) [inline AI/src/search/mcts/nodes/PNMCTSNode.java:290]: can remove this TODO comment, I've done it.

File: AI/src/search/mcts/nodes/PNMCTSNode.java [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+    	return (proofValue != PNMCTSNodeValues.UNKNOWN);
+    }
+    
+    @Override
+    public double expectedScore(final int agent)
+    {
+    	if (rootPlayer == agent)
+    	{
+    		if (proofValue == PNMCTSNodeValues.TRUE)
+    			return 1.0;
```
Comment (DennisSoemers) [inline AI/src/search/mcts/nodes/PNMCTSNode.java:331]: right now I'm only making the "expected score" (that is the score that things like Selection strategy and non-Robust final move selection strategies use) 1.0 if a win has been proven. If a node has been *dis*proven, I don't do anything, because I don't know if that means draw or loss.

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+            	
+            	if (currentNodeType == PNMCTSNodeTypes.AND_NODE)
+            	{
+            		// AND node
+            		for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: disproof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:177]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+            	}
+            	else
+            	{
+            		// OR node
+            		for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: proof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:194]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+                
+                if (currentNodeType == PNMCTSNodeTypes.AND_NODE)
+            	{
+            		// AND node
+                	for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: disproof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:243]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+            	}
+                else
+                {
+                	// OR node
+                	for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: proof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:261]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+                
+                if (sum > 0.0)
+                {
+                	for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						final double number;
+						if (child == null)
+						{
+							// This means: (dis)proof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:280]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+                
+                if (currentNodeType == PNMCTSNodeTypes.AND_NODE)
+            	{
+            		// AND node
+                	for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: disproof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:315]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+            	}
+                else
+                {
+                	// OR node
+                	for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: proof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:335]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+                
+                if (max > 0.0)
+                {
+                	for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						final double number;
+						if (child == null)
+						{
+							// This means: (dis)proof number = 1.0 for unexpanded child. TODO this correct?
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:359]: Jakub please check if this is what we want for children that have not been added to tree yet

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 17]
```diff
@@ -0,0 +1,429 @@
...
+	public void customise(final String[] inputs)
+	{
+		if (inputs.length > 1)
+		{
+			// We have more inputs than just the name of the strategy
+			for (int i = 1; i < inputs.length; ++i)
+			{
+				final String input = inputs[i];
+				
+				// TODO add other hyperparams
```
Comment (DennisSoemers) [inline AI/src/search/mcts/selection/PNS_UCB1.java:412]: I did not yet add nice code to allow customising all kinds of hyperparameters about the PN-part via constructors/strings. They do exist in the code directly.

File: AI/src/search/mcts/nodes/PNMCTSNode.java (hunk 4 above)
Comment (acatai) [reply AI/src/search/mcts/nodes/PNMCTSNode.java:128]: That probably depends, what we are proving. We had two ideas here, the simpler one was "I win against all others', so I assume this 1.0 then stays?
(The other one was with multiple trees, saying: mover wing against opponent `i`)

File: AI/src/search/mcts/nodes/PNMCTSNode.java (hunk 5 above)
Comment (acatai) [reply AI/src/search/mcts/nodes/PNMCTSNode.java:179]: I suppose so, it should work as a mobility heuristic that way..
Not sure is this disproof not too optimistic, but as a formula makes sense...

File: AI/src/search/mcts/nodes/PNMCTSNode.java (hunk 8 above)
Comment (acatai) [reply AI/src/search/mcts/nodes/PNMCTSNode.java:331]: Yep, correct.

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 18]
```diff
@@ -0,0 +1,429 @@
...
+	//-------------------------------------------------------------------------
+	
+	/** Exploration constant */
+	protected double explorationConstant;
+	
+	/** Constant by which to multiply the PN-based term */
+	protected double pnConstant;
+	
+	/** Minimum visits we must still allow sending into a child, even if it's a proven loss. */
+	protected int minVisitsSolvedChild = 5;
```
Comment (acatai) [inline AI/src/search/mcts/selection/PNS_UCB1.java:48]: that should be a param.
And even if not for now, then to turn things off it should be Integer.MAX

File: AI/src/search/mcts/selection/PNS_UCB1.java [hunk 19]
```diff
@@ -0,0 +1,429 @@
...
+                	continue;
+            }
+        	
+        	final double exploit;
+        	final double explore;
+
+        	if (child == null)
+        	{
+        		exploit = unvisitedValueEstimate;
+        		explore = Math.sqrt(parentLog);
```
Comment (acatai) [inline AI/src/search/mcts/selection/PNS_UCB1.java:115]: yy, it's not that we should skip normal selection in this case?
there is no need to pnsTerm, etc, we just get random unvisited child?

File: AI/src/search/mcts/selection/PNS_UCB1.java (hunk 9 above)
Comment (acatai) [reply AI/src/search/mcts/selection/PNS_UCB1.java:177]: aaa, OK, because we don't store the children values, so we get them call by need with ifology all the time.
Not sure if some encapsulation `getChildProof(possiblyNullNode)` would help make it cleaner.
Otherwise it is hopefully ok...

File: AI/src/search/mcts/MCTS.java (hunk 1 above)
Comment (acatai) [reply AI/src/search/mcts/MCTS.java:597]: It was not in my code, because the solver was using this 'skip' you implemented

File: AI/src/search/mcts/finalmoveselection/RobustChild.java (hunk 2 above)
Comment (acatai) [reply AI/src/search/mcts/finalmoveselection/RobustChild.java:20]: contempt will be for later, or very very later. For now we are concerned just by havin proven wins, so I suppose the quirks in the code are not of such importance (?).

File: AI/src/search/mcts/nodes/PNMCTSNode.java [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+     */
+    public double disproofNumber()
+    {
+    	return disproofNumber;
+    }
+    
+    /**
+     * @return Array of PNS-based terms for selection strategy for all of our children.
+     */
+    public double[] childrenPNSSelectionTerms()
```
Comment (acatai) [inline AI/src/search/mcts/nodes/PNMCTSNode.java:312]: woo, why?
we should never call for a selection term if ANY of our children is null
and a value here is based on proof/disproof of our children.
But if you want to store selection values of our children, they would require having pn/dpns of grandchildren?

Comment (acatai) [reply AI/src/search/mcts/nodes/PNMCTSNode.java:312]: nah, I'm wrong.
I mean never call if any child is null is correct - I vote for this.
But ofc we don't need grandchildren, only children pns/dpns.

File: AI/src/search/mcts/finalmoveselection/RobustChild.java (hunk 2 above)
Comment (DennisSoemers) [reply AI/src/search/mcts/finalmoveselection/RobustChild.java:20]: yeah, with proven wins the weirdness in my implementation won't be an issue, because such proven wins will never be beaten out by any higher score. Only if scores in between can be proven (which they can in multi-player games with Score Bounded), my code can currently be very weird.

File: AI/src/search/mcts/nodes/PNMCTSNode.java (hunk 4 above)
Comment (DennisSoemers) [reply AI/src/search/mcts/nodes/PNMCTSNode.java:128]: Not if you're already mid-game and some other player has already won. Like, a 6-player game in which first place has already been secured by some opponent, but the remaining 5 players are still competing over the other ranks. At that point in the game, getting second place would be "winning" (it'd be the best result that remains achievable), but it would have a utility score lower than 1.0.

Comment (acatai) [reply AI/src/search/mcts/nodes/PNMCTSNode.java:128]: True in the real sense. 
But Marrk told to count only wines for the result tables even in multiplayer game so 2/3 is still lose then :P.
I don't like this particularly but so far have no idea how to represent such results nicely.

Comment (DennisSoemers) [reply AI/src/search/mcts/nodes/PNMCTSNode.java:128]: If you're searching from a position (like, the initial game state) in which everything is still possible, sure. But if even in your root game state some of the top ranks have already been claimed, and only lower ranks are still up for grabs, I see no reason not to treat whichever is the best result still possible as a "win". Just, Ludii would give a number lower than 1.0 to it. But we should have code somewhere to ask Ludii "what is the highest utility that is still possible?"
